File: archer/cli/commands/list.py
import click

from archer.utils import discover_templates, load_template_safely
from archer.cli.help import get_list_help
from archer.constants import MODE_SINGLE


@click.command()
def list_templates() -> None:
    """List all available templates from the default templates directory.

    Displays all built-in templates with their mode indicators and descriptions.
    Use this command to discover available validation templates before validation.
    """
    
    templates = discover_templates()

    if not templates:
        click.echo("No templates found.")
        return

    click.echo(f"Available templates ({len(templates)}):")
    click.echo()

    for template_name in sorted(templates):
        template = load_template_safely(template_name)

        if template:
            mode_indicator = f"[{template.mode}]" if template.mode else f"[{MODE_SINGLE}]"
            click.echo(f"  {template.name:<15} {mode_indicator:<12} - {template.description}")
        else:
            click.echo(f"  {template_name:<15} {'[invalid]':<12} - [Invalid template]")

File: internal/cli/list.go
package cli

import (
	"fmt"
	"sort"

	"github.com/spf13/cobra"

	"github.com/theinfosecguy/archer/internal/constants"
	"github.com/theinfosecguy/archer/internal/templates"
)

var listCmd = &cobra.Command{
	Use:   "list",
	Short: "List all available templates",
	Long: `List all available templates from the default templates directory.

Displays all built-in templates with their mode indicators and descriptions.
Use this command to discover available validation templates before validation.`,
	RunE: runList,
}

func runList(cmd *cobra.Command, args []string) error {
	templateNames := templates.DiscoverTemplates(constants.DefaultTemplatesDir)

	if len(templateNames) == 0 {
		fmt.Println("No templates found.")
		return nil
	}

	fmt.Printf("Available templates (%d):\n\n", len(templateNames))

	loader := templates.NewTemplateLoader(constants.DefaultTemplatesDir)

	sort.Strings(templateNames)
	for _, name := range templateNames {
		template, err := loader.GetTemplate(name)
		if err != nil {
			fmt.Printf("  %-15s [%-10s] - %s\n", name, "invalid", "[Invalid template]")
			continue
		}

		mode := template.Mode
		if mode == "" {
			mode = constants.ModeSingle
		}

		fmt.Printf("  %-15s [%-10s] - %s\n", template.Name, mode, template.Description)
	}

	return nil
}
